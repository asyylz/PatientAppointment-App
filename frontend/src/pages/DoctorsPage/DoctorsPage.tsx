import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppDispatch } from './../../store/index';
import { fetchDoctors } from '../../store/doctors-slice/doctors-slice';
import DoctorProfilCard from '../../components/DoctorProfilCard/DoctorProfilCard';
import classes from './DoctorsPage.module.css';
import { doctorActions } from '../../store/doctors-slice/doctors-slice';
import useHttp from '../../hooks/useHttp/useHttp';
import PaginationButtons from '../../components/PaginationButtons/PaginationButtons';
import { useLocation } from 'react-router-dom';
import Loader from '../../components/Loader/Loader';

const Doctors: React.FC = () => {
  const dispatch: AppDispatch = useDispatch();
  const { getDoctorWithAvailabilities } = useHttp();
  const location = useLocation();

  const {
    entities: doctors,
    status,
    error,
  } = useSelector((state: RootState) => state.doctors);

  const searchWord = useSelector((state: RootState) => state.search);

  // we call doctor with availabilities
  const handleSelectDoctor = async (doctor: Doctor) => {
    const doctorWithAvailability = await getDoctorWithAvailabilities(
      doctor._id
    );
    dispatch(doctorActions.selectDoctor(doctorWithAvailability));
  };
  const [pagination, setPagination] = useState<number>(1);
  const [filteredDoctors, setFilteredDoctors] = useState(doctors);


  // Extract departmentId from query parameters
  const queryParams = new URLSearchParams(location.search);
  const departmentId = queryParams.get('departmentId');


  useEffect(() => {
    dispatch(
      fetchDoctors(
        `http://localhost:3000/api/v1/doctors?limit=2&page=${pagination}&sort=firstName${
          departmentId ? `&departmentId=${departmentId}` : ''
        }`
      )
    );
  }, [pagination, departmentId, dispatch]);

  useEffect(() => {
    if (searchWord) {
      const filtered = doctors.filter(
        (doctor: Doctor) =>
          doctor.firstName.toLowerCase().startsWith(searchWord.toLowerCase()) ||
          doctor.lastName.toLowerCase().startsWith(searchWord.toLowerCase())
      );
      setFilteredDoctors(filtered);
    } else {
      setFilteredDoctors(doctors);
    }
  }, [searchWord, doctors]);

  return (
    <>
      <p className={classes.header}>OUR DOCTORS</p>
      <hr />
      <div role="doctors-container" className={classes.container}>
        {status === 'loading' && <Loader />}

        {status === 'succeeded' &&
          filteredDoctors?.map((doctor: Doctor) => (
            <div key={doctor._id}>
              <DoctorProfilCard
                doctor={doctor as Doctor}
                onSelectDoctor={handleSelectDoctor}
              />
            </div>
          ))}
        {status === 'failed' && <p>{error}</p>}
        {doctors.length === 0 && status === 'succeeded' && (
          <p className={classes['no--available']}>No available doctor.</p>
        )}
      </div>
      <div className={classes['pagination__wrapper']}>
        <PaginationButtons
          pagination={pagination}
          setPagination={setPagination}
          length={doctors?.length}
          limit={2}
        />
      </div>
    </>
  );
};

export default Doctors;
