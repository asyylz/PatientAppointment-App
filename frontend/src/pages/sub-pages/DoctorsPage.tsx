import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppDispatch } from './../../store/index';
import { fetchDoctors } from './../../store/doctors-slice';
import DoctorProfilCard from '../../components/UI/DoctorProfilCard';
import classes from './DoctorsPage.module.css';
import { doctorActions } from './../../store/doctors-slice';

const Doctors: React.FC = () => {
  const dispatch: AppDispatch = useDispatch();

  const {
    entities: doctors,
    status,
    error,
  } = useSelector((state: RootState) => state.doctors);

  const searchWord = useSelector((state: RootState) => state.search);

  const handleSelectDoctor = (doctor: Doctor) => {
    dispatch(doctorActions.selectDoctor(doctor));
  };

  const [filteredDoctors, setFilteredDoctors] = useState(doctors);

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchDoctors());
    }
  }, [status, dispatch]);

  useEffect(() => {
    if (searchWord) {
      const filtered = doctors.filter(
        (doctor) =>
          doctor.firstName.toLowerCase().startsWith(searchWord.toLowerCase()) ||
          doctor.lastName.toLowerCase().startsWith(searchWord.toLowerCase())
      );
      // .sort((a, b) => {
      //   const aFirstnameIncludes = a.firstName
      //     .toLowerCase()
      //     .includes(searchWord.toLowerCase());

      //   const bFirstnameIncludes = b.firstName
      //     .toLowerCase()
      //     .includes(searchWord.toLowerCase());

      //   if (aFirstnameIncludes && !bFirstnameIncludes) {
      //     return -1; // `a` should come before `b`
      //   } else if (!aFirstnameIncludes && bFirstnameIncludes) {
      //     return 1; // `b` should come before `a`
      //   } else {
      //     return 0; // keep the order as is
      //   }
      // });

      setFilteredDoctors(filtered);
    } else {
      setFilteredDoctors(doctors);
    }
  }, [searchWord, doctors]);

  return (
    <>
      <h1 className={classes.header}>OUR DOCTORS</h1>
      <hr />
      <div
        //style={{border:'1px solid red'}}
        className={classes.container}
      >
        {status === 'loading' && <p>Loading...</p>}

        {status === 'succeeded' &&
          filteredDoctors.map((doctor: Doctor, index) => (
            <DoctorProfilCard
              key={index}
              doctor={doctor as Doctor}
              onSelectDoctor={handleSelectDoctor}
            />
          ))}
        {status === 'failed' && <p>{error}</p>}
      </div>
    </>
  );
};

export default Doctors;
