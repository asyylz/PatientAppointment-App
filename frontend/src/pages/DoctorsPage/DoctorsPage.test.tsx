import renderer from 'react-test-renderer';
import { act } from 'react';
import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { combineReducers } from 'redux';
import doctorsReducer from '../../store/doctors-slice/doctors-slice';
import { configureStore } from '@reduxjs/toolkit/react';
import searchReducer from '../../store/search-slice/search-slice.ts';
import {
  toContainRole,
  renderComponent,
} from './../../_testUtils/mocks/helper.tsx';

import DoctorsPage from './../DoctorsPage/DoctorsPage.tsx';

expect.extend({ toContainRole });

// Start store with initial state for `currentUser`
const initialCurrentUserState: CurrentUser = {
  status: 'login success',
  token:
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjY2NzM2NjJmYmQ0MmE5NjZiNzVkZWM5MiIsImlhdCI6MTcyMzc1NDAzOSwiZXhwIjoxNzIzNzU1ODM5fQ.4Nmyge8591WiGN3wEijyn-9Tx7A0GCBu5w2a0CgVaQs',
  userData: {
    _id: 'mocked_id',
    name: 'Asiye',
    email: 'alice@test.com',
    role: 'patient',
    __v: 0,
    DOB: '1986-01-22T00:00:00.000Z',
    image: 'mocked_image_url',
    policy: true,
  },
  error: null,
};

const mockAllReducers = combineReducers({
  doctors: doctorsReducer,
  search: searchReducer,
  currentUser: jest.fn(() => {
    return {
      ...initialCurrentUserState,
      // other state
    };
  }),
  // other reducers...
});

const store = configureStore({
  reducer: mockAllReducers,
});

describe('Doctors Page', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  /* -------------------------- - ------------------------- */
  it('1--Matches the DOM snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <DoctorsPage />
        </MemoryRouter>
      </Provider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  /* -------------------------- - ------------------------- */
  it('2--After rendering useEffect fetch doctors', async () => {
    renderComponent(
      <>
        <DoctorsPage />
      </>,
      store
    );

    expect(store.getState().doctors.status).toEqual('loading');
    // console.log(store.getState().doctors.status);
    expect(screen.getByRole('loader'));

    expect(
      store.dispatch({
        type: 'doctors/fetch/pending',
      })
    ).toHaveBeenCalled;

    expect(store.getState().doctors.status).toEqual('loading');
    console.log(store.getState().doctors.status);

    await act(async () => {
      store.dispatch({
        type: 'doctors/fetch/fulfilled',
        payload: [
          {
            _id: '665f0f2959c971659af920e5',
            gender: 'F',
            image: '/img/doctors/drmathai_small.jpg',
            phone: '647-722-2370',
            address: {
              street: '390 Steeles Avenue West',
              city: 'Vaughan',
              prov: 'ON',
              postal: 'L4J',
              _id: '665f0f2959c971659af920e7',
            },
            firstName: 'Aytekin Test',
            lastName: 'Yaliz',
            departmentId: {
              _id: '6660fbd12e3c00fe18cfceec',
              departmentMain: 'Allergy & Immunology',
            },
            userId: {
              _id: '666c1f10ad015e6df17a98e4',
              image:
                'https://patient-appointment-system.s3.eu-west-2.amazonaws.com/1721905296892-doctor.jpg',
            },
            reviews: [],
          },
          {
            _id: '665f0f2959c971659af920ca',
            id: 1,
            gender: 'M',
            image: '/img/doctors/chan_small.jpg',
            phone: '416-486-1956',
            address: {
              street: '473 Dupont Street',
              city: 'Toronto',
              prov: 'ON',
              postal: 'M6G 1Y6',
              _id: '665f0f2959c971659af920cc',
            },
            firstName: 'Bowen Test',
            lastName: 'Chan',
            departmentId: {
              _id: '6660fbd12e3c00fe18cfceef',
              departmentMain: 'Pediatrics',
            },
            doctorDescription:
              'Dr. Bowen Chan is board-certified in Pediatric Endocrinology and specializes in the diagnosis and management of hormonal disorders in children and adolescents. His areas of expertise include diabetes mellitus, growth disorders, and thyroid conditions.',
            reviews: [],
          },
        ],
      });
    });

    expect(store.getState().doctors.status).toEqual('succeeded');
    expect(screen.findByText('Aytekin Test'));
    const doctorsContainer = screen.getByRole('doctors-container');
    expect(doctorsContainer).toContainRole('doctor', 2);
    //const doctorprofilCards = screen.getAllByRole('doctor');
    //const link = within(doctorprofilCards[0]).getByRole('link');
  });

  /* -------------------------- - ------------------------- */
  it('--clicking make appointment link button should select doctor', async () => {
    // const doctorprofilCards = await screen.findAllByRole('doctor');
    // const link = within(doctorprofilCards[0]).getByRole('link');
    // console.log(link)
  });

  /* -------------------------- - ------------------------- */
  it('--Search term typed and doctors being filtered', async () => {
    renderComponent(
      <>
        <DoctorsPage />
      </>,
      store
    );

    console.log(store.getState().doctors.status); //succeeded
    //screen.debug();
    await act(async () => {
      store.dispatch({
        type: 'search/setSearch',
        payload: 'Aytekin',
      });
    });
  });
});
