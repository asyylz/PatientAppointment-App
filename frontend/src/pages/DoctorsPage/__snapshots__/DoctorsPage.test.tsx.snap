// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doctors Page 1--Matches the DOM snapshot 1`] = `
<DocumentFragment>
  <p
    class="header"
  >
    OUR DOCTORS
  </p>
  <hr />
  <div
    class="container"
    role="doctors-container"
  >
    <div
      class="loader__container"
      role="loader"
    >
       
      <div
        class="loader__ldsRipple"
      >
        <div />
        <div />
      </div>
    </div>
  </div>
  <div
    class="pagination__wrapper"
  >
    <div
      class="pagination__wrapper"
    >
      <div
        class="pagination__icon--wrapper"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            fill="none"
            points="7 2 17 12 7 22"
            stroke-width="2"
            transform="matrix(-1 0 0 1 24 0)"
          />
        </svg>
        <p>
          Prev
        </p>
      </div>
      <p
        class="pagination__page-number"
      >
        Page: 1
      </p>
      <div
        class="pagination__icon--wrapper passive"
      >
        <p>
          Next
        </p>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            fill="none"
            points="7 2 17 12 7 22"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
